### ISSUE TITLE
cv::countNonZero fails with UMat and non-OpenCL allocation (e.g. StdMatAllocator)

### ISSUE THREAD

**Issue by diablodale** [opencv/opencv#20542]
While debugging another PR, I discovered that `cv::countNonZero` fails when given a `UMat` that has any allocator other than `OpenCLAllocator`.  For example, if given a UMat created when opencl is disable/unavailable...then this function will fail.

#### System information (version)

- OpenCV => 4.5.3
- Operating System / Platform => all
- Compiler => all

#### Detailed description

https://github.com/opencv/opencv/blob/4d63a89fa6611a69c1999328375f39afde1b1696/modules/core/src/count_non_zero.dispatch.cpp#L134-L138

The bug above is in the boolean logic to choose to run/not the OCL version of the counter. Errant code only checks if the `InputArray` is a UMat. It is insufficient. It needs to also confirm that this UMat is allocated by OpenCL. It is required to do that check here so that if it isn't allocated by OpenCL, then it will not run the OCL counter and fallback to code further down.

The fix is generally easy. Just add a test similar to `u->currAllocator == getOpenCLAllocator()`
However, I am also concerned with the current code having `OCL_PERFORMANCE_CHECK(_src.isUMat())`. It is essential to always in all cases check for both UMat and the Allocator. It isn't a choice of which path is faster/bypass. Instead, one path is impossible. So that macro seems improper in this situation.

#### Steps to reproduce

I found it when running the standard OpenCV test case `Core UMat.setOpenCL` on my laptop with...
```
OPENCV_OPENCL_DEVICE=NVIDIA:GPU:
OPENCV_OPENCL_SVM_FORCE_UMAT_USAGE=TRUE
```

1. `opencv_test_cored.exe --gtest_filter=UMat.*`

#### Result

```
[ RUN      ] UMat.setOpenCL
[ERROR:0] global ..\modules\core\src\ocl.cpp (3966) cv::ocl::Kernel::set OpenCL: Kernel(reduce)::set(arg_index=0, flags=258): can't create cl_mem handle for passed UMat buffer (addr=000000318CCFDE80)
```

#### Expected

No error. Test case to complete successfully.

#### Issue submission checklist

 - [X] I report the issue, it's not a question
 - [X] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
 - [X] I updated to latest OpenCV version and the issue is still there
 - [X] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by alalek** [comment]
That you for the report!

The error message is added after that test.
AFAIK, function code path should recover after this error. Test doesn't fail.

Perhaps we lost somewhere the code path when CPU buffer is wrapped in OpenCL buffer (e.g., some OpenCL devices have uniform memory with host - so there is no data copy) and then executed through OpenCL kernel. There is ACCESS_FAST designed for that.

**Comment by diablodale** [comment]
Test does fail in all my local testing. Just have to have the right setup as described above.

I have not seen any meaningful code that implements `ACCESS_FAST` in OpenCL. It seems to be a feature not used much. I see the enum appear in some OpenCL allocator code, but it isn't something that I'm investigating in my current PR. It is big enough without adding that on. 😉

In OpenCL, a feature that is similar to what you describe is SVM with three versions of it. Fine grain system is the 3rd and most system wide single address space with no mapping.

**Comment by alalek** [comment]
This is not SVM-only feature. `clCreateBuffer(CL_MEM_USE_HOST_PTR)` should also work for some cases.

BTW, `ACCESS_FAST` is designed, but probably it is not implemented well.

**Comment by diablodale** [comment]
No worries. I understand the OpenCl allocator now. I've put a lot of time into documenting the code, which helped me understand it. I intend to do several PRs to fill in missing code, fix issues, etc. `HOST_PTR` is a feature that I want to explore in a later PR because I also want to use that to enable SVM for `Image2D`.

Meanwhile, I'm in bug fix mode for my related device/context switching PR. Now that it most works, it exposes bugs in other OpenCV code so I need to fix these little bugs in other locations so I can get through a full test pass. 👍🚧🤪

**Comment by diablodale** [comment]
I've discovered a broader problem. 😢 This specific issue in `cv::countNonZero` is just one example of the same problem in 88+ other locations. Throughout the OpenCV codebase is the same errant code. Here are three examples from many... 

https://github.com/opencv/opencv/blob/0f8efb07c78441fa483ae6366fee8ff1a0b37286/modules/core/src/matrix_transform.cpp#L728

https://github.com/opencv/opencv/blob/0f8efb07c78441fa483ae6366fee8ff1a0b37286/modules/imgproc/src/blend.cpp#L388-L389

https://github.com/opencv/opencv/blob/0f8efb07c78441fa483ae6366fee8ff1a0b37286/modules/core/src/dxt.cpp#L3492-L3501

The problem in all these locations is the same. The errant code is assuming that when `isUMat() = true` that means the UMat is an OpenCLAllocator UMat. Unfortunately, this is not true for many cases. The easiest case to show is in the OP above and anytime that code creates a UMat when OpenCL is disabled (like in many test cases).

Since this errant logic is throughout the OpenCV codebase, I recommend a broad fix that will work everywhere. Here are some ideas.

1. Change the code in all 88+ locations to correctly check for UMat and for OpenCLAllocator. Like the fix proposed for the OP.
   This approach will fix the problem now. And relies on the developer to always remember to always test for both.
2. Change the Input/OutputArray::isUMat() api. Make it have a required parameter that forces the developer to indicate what
   type of UMat they are testing. For example `::isUMat(ANY)`, `::isUMat(OCL)`, `::isUMat(OCL_SVM)`, `isUMat(OCL_BUF)`, etc.
   Note, the latter has already been needed and coded with the `internal::isCLBuffer()` hack. This approach forces the developer
   to always think about what they need.
3. Change the Input/OutputArray::isUMat() api. Make it have the same parameter but with a default value of `ANY`. This
   provides the infrastructure/api for umat allocator checking code to be centralized and usable. Existing code will compile
   but will be errant in these 88+ locations. So we still need to correct that code to use `::isUMat(OCL)`. Devs can make the
   same mistake again, but better test cases in the CI should catch them + now they have an api for a better OCL_RUN check.
4. Others??

**Comment by alalek** [comment]
`isUMat()` is part of dispatcher check. We want to keep dispatching rules simple.
isCLBuffer doesn't make a lot of sense here, because as I wrote above CPU memory can be wrapped into cl_buffer at zero cost on some configurations.
Perhaps code parts with `getUMat` calls should be updated instead (as they work with OpenCL buffers).

BTW, There are were several ideas to make `UMat` as really universal:
- currently it works with cl_buffer and forces its creation, even if some part of pipeline updates data on CPU only and don't call OpenCL kernels.
- recently we receive multiple `OpenCLExecutionContext`s. OpenCL buffer can belong to another context than the current context attached to thread.
- there were some discussions to wrap more data kinds in UMat, e.g., CUDA handles, VA buffers, DirectX textures, etc.
- unfortunately efficient dispatcher requires knowing a whole pipeline to minimize possible non-zero costs of data transferring.

**Comment by diablodale** [comment]
changing `getUMat()` I don't think helps us. `getUMat()` should continue to get a UMat using the currently bound execution context (which contains a pointer to the allocator). I do not think that should change.

Unfortunately, the RUN_OCV dispatcher code in these ~100 places is insufficient. It is not reliable as is -- because the logic isn't distinct enough to be reliable. There are not enough boolean tests to clearly check for requirements. Not all UMats support OpenCL. And therefore it is not proper to call a function that *only* works on OpenCL if the code only checks for UMat.

It will work when OpenCL is not compiled, or disabled. That is why all the test cases are passing on most CI platforms for years.
There are only two CI platform that compile for OpenCL. And on those two there is no SVM. And no context switching. So a lot of test cases are not being tested. And I've fixed some bugs. That's just history and good evolution both. Your team has to prioritize time/effort. 👍

I can't think of any way to automatically guess what a dev wants. Sometimes it is just a UMat w/ OpenCLAllocator. Other times, like the gemm_amd thing, the dev specifically wants a UMat w/ OpenCLAllocator w/ cl_mem (because the AMD code doesn't support SVM). Also, consider the API itself `.isUMat()`. That API is doing exactly what is declares. It tells the dev "is this a UMat or not a UMat". That API declares nothing about allocators, buffers, buffer types, etc. To determine that, a changed/new/different API has to be used. That is why you tried to make `internal::isCLBuffer(UMat& u)` in 2015. You were trying to fix this issue of the OCL dispatcher not having enough information in it. But even that `internal::isCLBuffer(UMat& u)` function had a bug in it. And I fixed it in my PR.  ;-)

My focus is to unblock the work I've done for context switching. While coding it, I've fixed other issues which were needed for reliability. Now SVM is substantially better. And I'm running tests with compiles and platform settings that you don't test in your CI. This has exposed bugs in other areas like this issue. Again, this is not worrying and I'm not mad or disappointed. All is ok. We only need to come to some joint agreements and then I can start work on the fixes so I can unblock my larger work.

I would prefer to do minimal changes, and evolve the api rather than a complete rewrite. I'm also considering API compatibility with 4.x. I don't want to break it. So I am not leaning for a newer option #4.

4. Leave `isUMat()` as is. I can't change it...ABI compatibility. So I'll make a sibling API with a required parameter `isUMat(enum something)` which a dev can use to specify what type of UMat do they want to test. I know the code that's needed for that and I can leverage that code from places it is already written. Then I'll update all those ~100 places to use the appropriate enum for that API. Ny default I will use `isUMat(opencl)` because that is the logic that you and other devs assumed it did. That opencl param will cause the logic to check for something allocated by the OpenCL allocator (which currently is either cl_mem or svm pointer). I would consider just having the dev check the umat.u.currallocator themselves...but the allocator won't tell them if it is (cl_mem or svm). That level of specific distinction is needed for things like `gemm_amd` which doesn't support SVM. Luckily, anything that uses the ocl::Kernel apis now (with my PR) support both cl_mem and SVM. So it is a small set (maybe only gemm_amd) that need the !SVM check.

If I don't change those ~100 places, then my PR will not pass tests in certain configurations. Not compiled openCL or disabled OpenCL will work. And *some* scenarios of OpenCL with all default OpenCL settings will likely be ok too. But tests that turn opencl on/off or change contexts will probably fail.

So I propose option 4.  Do you have another specific idea?
